core/opendaq/functionblock/include/opendaq/function_block_type_impl.h

```diff
@@ -38,6 +40,105 @@ class FunctionBlockTypeImpl : public GenericComponentTypeImpl<IFunctionBlockType
+{
+}
+
+inline ErrCode FunctionBlockTypeImpl::serialize(ISerializer* serializer)
+{
+    OPENDAQ_PARAM_NOT_NULL(serializer);
+
+    const auto serializerPtr = SerializerPtr::Borrow(serializer);
+
+    return daqTry(
+        [this, &serializerPtr]
+        {
+            serializerPtr.startTaggedObject(borrowPtr<SerializablePtr>());
+            {
+                serializerPtr.key("id");
+                serializerPtr.writeString(id);
+
+                if (name.assigned())
+                {
+                    serializerPtr.key("name");
+                    serializerPtr.writeString(name);
+                }
+
+                if (description.assigned())
+                {
+                    serializerPtr.key("description");
+                    serializerPtr.writeString(description);
+                }
+
+                if (createDefaultConfigCallback.assigned())
+                {
+                    const auto defaultConfig = createDefaultConfigCallback.call();
+                    if (defaultConfig.assigned())
+                    {
+                        serializerPtr.key("defaultConfig");
+                        defaultConfig.serialize(serializerPtr);
+                    }
+                }
+            }
+
+            serializerPtr.endObject();
+        });
+}
+
+inline ErrCode FunctionBlockTypeImpl::getSerializeId(ConstCharPtr* id) const
+{
+    OPENDAQ_PARAM_NOT_NULL(id);
+
+    *id = SerializeId();
+    return OPENDAQ_SUCCESS;
+}
+
+inline ConstCharPtr FunctionBlockTypeImpl::SerializeId()
+{
+    return "FunctionBlockType";
+}
+
+inline ErrCode FunctionBlockTypeImpl::Deserialize(ISerializedObject* serialized,
+                                                  IBaseObject* context,
+                                                  IFunction* factoryCallback,
+                                                  IBaseObject** obj)
+{
+    OPENDAQ_PARAM_NOT_NULL(serialized);
+    OPENDAQ_PARAM_NOT_NULL(obj);
+
+    const auto serializedObj = SerializedObjectPtr::Borrow(serialized);
+    const auto contextPtr = BaseObjectPtr::Borrow(context);
+    const auto factoryCallbackPtr = FunctionPtr::Borrow(factoryCallback);
+
+    return daqTry(
+        [&serializedObj, &contextPtr, &factoryCallbackPtr, &obj]
+        {
+            const auto id = serializedObj.readString("id");
+
+            StringPtr name;
+            if (serializedObj.hasKey("name"))
+                name = serializedObj.readString("name");
+
+            StringPtr description;
+            if (serializedObj.hasKey("description"))
+                description = serializedObj.readString("description");
+
+            FunctionPtr createDefaultConfig;
+            if (serializedObj.hasKey("defaultConfig"))
+            {
+                PropertyObjectPtr defaultConfig = serializedObj.readObject("defaultConfig", contextPtr, factoryCallbackPtr);
+                createDefaultConfig = Function([defaultConfig] { return defaultConfig; });
+            }
```

Comment:
Not sure if it's really correct to deserialize the object in this manner. Also, I think that `defaultConfig` returned by the lambda here will be nullptr, as the property object ptr will go out of scope.

I don't really think there's a good way of implementing this if we don't have a `clone` method for the property object.

In general, this is also missing a test case.

Reply:
`defaultConfig` is passed to lambda by value, so the ref count should be incremented. Since this code is just moved from cpp to h, I would spend much time on it (at least not under the scope of this ticket).